crates/bevy_render/src/pipeline_keys/composite.rs

```diff
@@ -0,0 +1,115 @@
+    ($(($K:ident, $sz:ident, $value:ident)),*) => {
+        impl<$($K: PipelineKeyType),*> PipelineKeyType for ($($K,)*) {
+            fn positions(store: &KeyMetaStore) -> HashMap<TypeId, SizeOffset> {
+                let mut result = HashMap::default();
+                let mut offset = 0;
+
+                $(
+                    let $sz = $K::size(store);
+                    offset += $sz;
+                )*
+
+                $(
+                    offset -= $sz;
+                    result.insert(TypeId::of::<$K>(), SizeOffset{ size: $sz, offset });
+                )*
+
+                result
+            }
+
+            fn pack(value: &Self, store: &KeyMetaStore) -> PackedPipelineKey<Self> {
+                let mut result = 0 as KeyPrimitive;
+                let mut total_size = 0u8;
+
+                let ($($value,)*) = value;
+
+                $(
+                    let PackedPipelineKey{ packed, size, .. } = $K::pack($value, store);
+                    assert_eq!(size, $K::size(store), "{} mismatch", type_name::<$K>());
+
+                    result = (result << size) | packed;
+                    total_size += size;
+                    debug!("{} added {}={:#b} @ {} -> {:#b}", type_name::<Self>(), type_name::<$K>(), packed, size, result);
+                )*
+
+                PackedPipelineKey::new(result, total_size)
+            }
+
+            fn unpack(value: KeyPrimitive, store: &KeyMetaStore) -> Self {
+                let mut shift_bits = 0;
+                $(
+                    let $sz = $K::size(store);
+                    shift_bits += $sz;
+                )*
+
+                (
+                    $({
+                        shift_bits -= $sz;
+                        $K::unpack((value >> shift_bits) & ((1 << $sz) - 1), store)
+                    },)*
+                )
+            }
+        }
+
+        impl<$($K: PipelineKeyType),*> CompositeKey for ($($K,)*) {
+            fn from_keys(keys: &PipelineKeys) -> Option<PackedPipelineKey<Self>> {
+                let mut result = 0 as KeyPrimitive;
+                let mut total_size = 0u8;
+
+                $(
+                    let PackedPipelineKey{ packed, size, .. } = keys.get_packed_key::<$K>()?;
+                    result = result << size | packed;
+                    total_size += size;
+                )*
+
+                Some(PackedPipelineKey::new(result, total_size))
+            }
+
+            fn set_config() -> NodeConfigs<Interned<dyn bevy_ecs::schedule::SystemSet>> {
+                let mut config = KeySetMarker::<Self>::default().after(KeySetMarker::<()>::default());
+
+                $(
+                    config = config.after(KeySetMarker::<$K>::default());
+                )*
+
+                config
+            }
+        }
+
+        impl<$($K: FixedSizeKey),*> FixedSizeKey for ($($K,)*) {
+            fn fixed_size() -> u8 {
+                $($K::fixed_size() + )* 0
```

Comment:
This works I guess, but isn't there some kind of "put delimiter in between but not after the last one" feature in macros?

If not, you could emit a loop that adds all the fixed size values together, like you did below in `KeyRepack`.

Reply:
if it was a comma, then `$($x),*` would do commas inbetween the elements and `$(x,)*` would do a trailing comma as well.

unfortunately because `+` is also a "MacroRepOp" (`+`, `*`, `?`) it can't be used as a "MacroRepSep" (anything except a MacroRepOp). i have no idea what `+` or `?` actually do in that position though.

i could do a loop if this is too awkward ... though i think most people look at the expansion, which is pretty clear.